### ISSUE TITLE
[BUG] Volume issues

### ISSUE THREAD

**Issue by Shardj** [Crylia/crylia-theme#24]
Trying to change my volume via the keyboard resulted in 'ERROR attempt to compare number with nil'. I soon noticed from the top bar's volume button that I don't have an output device selected which is likely the problem. However trying to select the only output device in the list resulted in 'ERROR .config/awesome/src/modules/volume_controller.lua:58: attempt to concatenate a nil value (upvalue 'node')

**Comment by Crylia** [comment]
Do you have pactl (pulseaudio) installed?

**Comment by Shardj** [comment]
Yes

**Comment by Crylia** [comment]
does `pactl list sinks`  work? And if yes, can you please post the output?

**Comment by Shardj** [comment]
```
Sink #0
        State: SUSPENDED
        Name: alsa_output.pci-0000_00_1f.3.analog-stereo
        Description: Built-in Audio Analogue Stereo
        Driver: module-alsa-card.c
        Sample Specification: s16le 2ch 48000Hz
        Channel Map: front-left,front-right
        Owner Module: 7
        Mute: yes
        Volume: front-left: 0 /   0% / -inf dB,   front-right: 0 /   0% / -inf dB
                balance 0.00
        Base Volume: 65536 / 100% / 0.00 dB
        Monitor Source: alsa_output.pci-0000_00_1f.3.analog-stereo.monitor
        Latency: 0 usec, configured 0 usec
        Flags: HARDWARE HW_MUTE_CTRL HW_VOLUME_CTRL DECIBEL_VOLUME LATENCY
        Properties:
                alsa.resolution_bits = "16"
                device.api = "alsa"
                device.class = "sound"
                alsa.class = "generic"
                alsa.subclass = "generic-mix"
                alsa.name = "ALC289 Analog"
                alsa.id = "ALC289 Analog"
                alsa.subdevice = "0"
                alsa.subdevice_name = "subdevice #0"
                alsa.device = "0"
                alsa.card = "0"
                alsa.card_name = "HDA Intel PCH"
                alsa.long_card_name = "HDA Intel PCH at 0x6165298000 irq 204"
                alsa.driver_name = "snd_hda_intel"
                device.bus_path = "pci-0000:00:1f.3"
                sysfs.path = "/devices/pci0000:00/0000:00:1f.3/sound/card0"
                device.bus = "pci"
                device.vendor.id = "8086"
                device.vendor.name = "Intel Corporation"
                device.product.id = "43c8"
                device.form_factor = "internal"
                device.string = "front:0"
                device.buffering.buffer_size = "384000"
                device.buffering.fragment_size = "192000"
                device.access_mode = "mmap+timer"
                device.profile.name = "analog-stereo"
                device.profile.description = "Analogue Stereo"
                device.description = "Built-in Audio Analogue Stereo"
                module-udev-detect.discovered = "1"
                device.icon_name = "audio-card-pci"
        Ports:
                analog-output-speaker: Speakers (priority: 10000)
                analog-output-headphones: Headphones (priority: 9900, not available)
        Active Port: analog-output-speaker
        Formats:
                pcm
```

**Comment by Crylia** [comment]
It seems like your audio controller doesn't report a node.name. This is really weird and I wonder if it also happens to headphones?
I am probably going to switch to lgi and use the dbus instead of pactl so I won't spend more time fixing this.

Can you post what `pactl get-default-sink` does? with no node.name I'm wondering what it reports

**Comment by Shardj** [comment]
Just spits out "No valid command specified."

**Comment by Crylia** [comment]
> Just spits out "No valid command specified."

I think your pactl is broken or something? For me it reports the current sink.

**Comment by Shardj** [comment]
No, it works fine. Probably just different versions. `pactl info` gets me the default just fine:

Server String: /run/user/1000/pulse/native
Library Protocol Version: 33
Server Protocol Version: 33
Is Local: yes
Client Index: 215
Tile Size: 65472
User Name: george
Host Name: george-laptop
Server Name: pulseaudio
Server Version: 13.99.1
Default Sample Specification: s16le 2ch 44100Hz
Default Channel Map: front-left,front-right
Default Sink: alsa_output.pci-0000_00_1f.3.analog-stereo
Default Source: alsa_input.pci-0000_00_1f.3.analog-stereo
Cookie: e8f2:b78c

**Comment by Shardj** [comment]
Yeah just checked, the get-default-sink command is part of pulseaudio v15+ and I'm on v13.99.1

**Comment by Crylia** [comment]
Does it work once you update?

**Comment by Shardj** [comment]
13.99.1 is the latest PPA for pulse audio on ubuntu. I don't want to go and mess around with trying to upgrade it myself.

**Comment by Crylia** [comment]
I guess pulseaudio-utils or pipewire-pulse uses a higher version then because I'm running my stock repos from 22.04 Pop!_OS

**Comment by Shardj** [comment]
Can you use this snippet instead or add in a check for the get-default command failing and use this if so?

`pactl info | perl -n -e'/Default Sink: (.+)\s/ && print $1'`

**Comment by Crylia** [comment]
Yes I will add it

**Comment by Crylia** [comment]
Did that fix the problems you had?

**Comment by Shardj** [comment]
No unfortunately not

volume_controller line 58 attempt to concatenate a nil value (upvalue 'node')

**Comment by isaackhlam** [comment]
Hi, I also encounter the same problem, My laptop is rather new(ASUS UM3402YA), the audio is currently unavailable. I am using the arch disro to install the theme. And my speaker work at native Debian with KDE installed. (Decide to change cuz I want latest kernel). I followed existing discuss to print out the info:

`pactl list sinks`
```
Sink #0
	State: SUSPENDED
	Name: alsa_output.pci-0000_04_00.6.HiFi_hw_Generic_1_sink
	Description: Family 17h/19h HD Audio Controller Speaker + Headphones
	Driver: module-alsa-card.c
	Sample Specification: s16le 2ch 48000Hz
	Channel Map: front-left,front-right
	Owner Module: 8
	Mute: yes
	Volume: front-left: 13762 /  21% / -40.67 dB,   front-right: 13762 /  21% / -40.67 dB
	        balance 0.00
	Base Volume: 65536 / 100% / 0.00 dB
	Monitor Source: alsa_output.pci-0000_04_00.6.HiFi_hw_Generic_1_sink.monitor
	Latency: 0 usec, configured 0 usec
	Flags: HARDWARE HW_MUTE_CTRL HW_VOLUME_CTRL DECIBEL_VOLUME LATENCY
	Properties:
		alsa.resolution_bits = "16"
		device.api = "alsa"
		device.class = "sound"
		alsa.class = "generic"
		alsa.subclass = "generic-mix"
		alsa.name = "ALC294 Analog"
		alsa.id = "ALC294 Analog"
		alsa.subdevice = "0"
		alsa.subdevice_name = "subdevice #0"
		alsa.device = "0"
		alsa.card = "1"
		alsa.card_name = "HD-Audio Generic"
		alsa.long_card_name = "HD-Audio Generic at 0xfccc0000 irq 84"
		alsa.driver_name = "snd_hda_intel"
		device.bus_path = "pci-0000:04:00.6"
		sysfs.path = "/devices/pci0000:00/0000:00:08.1/0000:04:00.6/sound/card1"
		device.bus = "pci"
		device.vendor.id = "1022"
		device.vendor.name = "Advanced Micro Devices, Inc. [AMD]"
		device.product.id = "15e3"
		device.product.name = "Family 17h/19h HD Audio Controller"
		device.string = "_ucm0003.hw:Generic_1"
		device.buffering.buffer_size = "352768"
		device.buffering.fragment_size = "176384"
		device.access_mode = "mmap+timer"
		device.profile.name = "HiFi: hw:Generic_1: sink"
		device.profile.description = "Speaker + Headphones"
		alsa.mixer_device = "_ucm0003.hw:Generic_1"
		device.description = "Family 17h/19h HD Audio Controller Speaker + Headphones"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card-pci"
	Ports:
		[Out] Speaker: Speaker (type: Speaker, priority: 100, availability unknown)
		[Out] Headphones: Headphones (type: Headphones, priority: 200, not available)
	Active Port: [Out] Speaker
	Formats:
		pcm
```
`pactl get-default-sink`
```
alsa_output.pci-0000_04_00.6.HiFi_hw_Generic_1_sink
```
`pactl --version`

```
pactl 16.1
Compiled with libpulse 16.1.0
Linked with libpulse 16.1.0
```
When I click the audio widget, it also say`volume_controller line 58 attempt to concatenate a nil value (upvalue 'node')`

**Comment by isaackhlam** [comment]
Update: It might my driver issue, I found that some one else encounter speaker issue from [stackExchange](https://superuser.com/questions/1719920/no-sound-from-internal-speakers-on-laptop-but-headphone-jack-and-hdmi-works).
Interestingly, headphones work properly.

**Comment by Crylia** [comment]
The problem is that neither of you got a `node.name` which I need in my current setup to set/change the current sink/source.
I don't think I can really change this without switching to https://github.com/sclu1034/lua-libpulse-glib.

**Comment by jiangqingyuan** [comment]
I switched hdmi to speaker using gnome,  and then login awesomewm, it was okay.

**Comment by sharedordaz** [comment]
I Got the  same problem
![Volume-error](https://github.com/Crylia/crylia-theme/assets/82840030/7e718b34-8385-4a10-b347-886c79479ee2)

**Comment by jb1277976** [comment]
Pinging. Did this bug ever get resolved ? I was thinking of trying this theme Desktop but just looking over everything first!

**Comment by Crylia** [comment]
> Pinging. Did this bug ever get resolved ? I was thinking of trying this theme Desktop but just looking over everything first!

Well yeah in the dev branch, I planned on adding a shit ton of stuff and rewriting most things but never got to finish it and I'm not using awesome anymore. If anyone wants to merge something feel free to open a PR but I probably won't make anything in the near future

**Comment by jb1277976** [comment]
> > Pinging. Did this bug ever get resolved ? I was thinking of trying this theme Desktop but just looking over everything first!
> 
> Well yeah in the dev branch, I planned on adding a shit ton of stuff and rewriting most things but never got to finish it and I'm not using awesome anymore. If anyone wants to merge something feel free to open a PR but I probably won't make anything in the near future

What window manager do you use these days ? i'm curious!

**Comment by Crylia** [comment]
> > > Pinging. Did this bug ever get resolved ? I was thinking of trying this theme Desktop but just looking over everything first!
> > 
> > 
> > Well yeah in the dev branch, I planned on adding a shit ton of stuff and rewriting most things but never got to finish it and I'm not using awesome anymore. If anyone wants to merge something feel free to open a PR but I probably won't make anything in the near future
> 
> What window manager do you use these days ? i'm curious!

KDE, mostly because of wallpaper engine and It's easy to theme, it just works and gnome breaks too often. I'm looking forward to CosmicDE from pop os

**Comment by jb1277976** [comment]
> > > > Pinging. Did this bug ever get resolved ? I was thinking of trying this theme Desktop but just looking over everything first!
> > > 
> > > 
> > > Well yeah in the dev branch, I planned on adding a shit ton of stuff and rewriting most things but never got to finish it and I'm not using awesome anymore. If anyone wants to merge something feel free to open a PR but I probably won't make anything in the near future
> > 
> > 
> > What window manager do you use these days ? i'm curious!
> 
> KDE, mostly because of wallpaper engine and It's easy to theme, it just works and gnome breaks too often. I'm looking forward to CosmicDE from pop os

I just stopped using awesome today and went back to i3 I'm going to learn it and not just copy stuff lol
